### PR TITLE
[`pycodestyle`] Handle f-strings properly for `invalid-escape-sequence (W605)`

### LINKED ISSUE #11491: Wrong fix for W605 in f-strings
Python: 3.12.3
Version: ruff 0.4.4
Command: `ruff check --select W605 .\test.py --fix`

And let's given the following code snippet in `test.py`:

```
total = 10
ok = 7
incomplete = 3
s = f"TOTAL: {total}\nOK: {ok}\INCOMPLETE: {incomplete}\n"
```

It is wrongly fixed as:

```
total = 10
ok = 7
incomplete = 3
s = rf"TOTAL: {total}\nOK: {ok}\INCOMPLETE: {incomplete}\n"
```

Now the string is a raw string. The newlines will then interpreted as literals and will not remain newlines.

---

If I do the same for a normal string:

```
s = "TOTAL: 10\nOK: 3\INCOMPLETE: 7\n"
```

It is correctly fixed as:

```
s = "TOTAL: 10\nOK: 3\\INCOMPLETE: 7\n"
```

The invalid escape sequence '\I' is correctly fixed as `\\`, and the newlines are still there. 

---

Even if it is a f-string, but without placeholders, it is done correctly:

```
s = f"TOTAL: 10\nOK: 3\INCOMPLETE: 7\n"
```

Result is correct:

```
s = f"TOTAL: 10\nOK: 3\\INCOMPLETE: 7\n"
```

### PR DESCRIPTION
**Pull request by dylwil3**
When fixing an invalid escape sequence in an f-string, each f-string element is analyzed for valid escape characters prior to creating the diagnostic and fix. This allows us to safely prefix with `r` to create a raw string if no valid escape characters were found anywhere in the f-string, and otherwise insert backslashes.

This fixes a bug in the original implementation: each "f-string part" was treated separately, so it was not possible to tell whether a valid escape character was or would be used elsewhere in the f-string.

Progress towards #11491 but format specifiers are not handled in this PR.

### CODE REVIEW

File: crates/ruff_linter/src/rules/pycodestyle/rules/invalid_escape_sequence.rs [hunk 1]
```diff
@@ -225,7 +252,26 @@ fn check(
...
+    diagnostics: &mut Vec<Diagnostic>,
+    locator: &Locator,
+    // Start position of the expression that contains the source range. This is used to generate
+    // the fix when the source range is part of the expression like in f-string which contains
+    // other f-string literal elements.
+    expr_start: TextSize,
+    flags: AnyStringFlags,
+    escape_chars_state: EscapeCharsState,
+) {
+    let EscapeCharsState {
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/pycodestyle/rules/invalid_escape_sequence.rs:271]: Nit: The method name is somewhat long and doesn't tell me much. I'd probably just inline the code into `check`, considering that `check` only contains one additional function call

Comment (dylwil3) [reply crates/ruff_linter/src/rules/pycodestyle/rules/invalid_escape_sequence.rs:271]: Happy to change the method name!

I don't think I can inline this into `check` though. Right now `check` is the composite of `analyze_escape_chars` and `check_from_escape_chars_state`. For ordinary strings we can just do `check` all at once, but the purpose of the refactoring was that, for f-strings, I need to do `analyze_escape_chars` on each f-string part _before_ doing `check_from_escape_chars_state`. 

Sorry if I'm misunderstanding the suggestion!

I'll try to add some comments around this and think of a better name.

Comment (MichaReiser) [reply crates/ruff_linter/src/rules/pycodestyle/rules/invalid_escape_sequence.rs:271]: Ah I see. I'd be inclined to inline `check` into the string literal and byte literal branches (it's really just that one `analyze` call). Actually, I'd probably rewrite it to 


```rust
				let state = match part {
            StringLikePart::String(_) | StringLikePart::Bytes(_) => {
							analyze_escape_chars(locator, part.range(), part.flags())
            }
            StringLikePart::FString(f_string) => {
                let flags = AnyStringFlags::from(f_string.flags);
                let mut escape_chars_state = EscapeCharsState::default();
                for element in &f_string.elements {
                    match element {
                        FStringElement::Literal(literal) => {
                            escape_chars_state.update(analyze_escape_chars(
                                locator,
                                literal.range(),
                                flags,
                            ));
                        }
                        FStringElement::Expression(expression) => {
                            let Some(format_spec) = expression.format_spec.as_ref() else {
                                continue;
                            };
                            for literal in format_spec.elements.literals() {
                                escape_chars_state.update(analyze_escape_chars(
                                    locator,
                                    literal.range(),
                                    flags,
                                ));
                            }
                        }
                    }
                }
                check_from_escape_chars_state(
                    &mut checker.diagnostics,
                    locator,
                    f_string.start(),
                    flags,
                    escape_chars_state,
                );
            }
				};

        check(
            &mut checker.diagnostics,
            locator,
            part.start(),
            part.range(),
            part.flags(),
						state
        );

```

Comment (MichaReiser) [reply crates/ruff_linter/src/rules/pycodestyle/rules/invalid_escape_sequence.rs:271]: Finger crossed that part has all these methods

Comment (dylwil3) [reply crates/ruff_linter/src/rules/pycodestyle/rules/invalid_escape_sequence.rs]: perfect, much better!

**Review by MichaReiser — APPROVED** [review]

**Comment by dylwil3** [comment]
> Do we need to have any special handling for escapes in format specifiers or the debug text parts of an f-string? We don't have to fix this as part of this PR but format specifiers and debug texts were something that caused me a a lot of pain when working on f-string formatting

As far as I can tell you can't really put an escape in either of those places: for format specifiers I think that's a [syntax error](https://docs.python.org/3/library/string.html#format-specification-mini-language) and for debug text I think you could only do that inside a nested string (e.g. `f"{'\InHere'=}"`), which is handled fine (added test to fixture).

But maybe I missed something so if someone comes up with an example I can do a followup PR.

Thank you for the thoughtful review!

**Comment by dylwil3** [comment]
> What you linked to is only the list of standard format specifiers but an object can implement support for arbitrary format specifiers

Terrifying.

Yes, this doesn't seem quite right when I make the escape invalid:

```console
➜  cargo run -p ruff -- check tmp.py --no-cache --select W605 --fix --diff
--- tmp.py
+++ tmp.py
@@ -1,2 +1,2 @@
 a = 10
-f"{a:\Xff}"
+rf"{a:\Xff}"

Would fix 1 error.
```
I'll leave the linked issue open and do a followup PR once I understand better how these things work...

Thank you!

**Pull request merged**